psbt: Taproot fields for PSBT

achow101 opened the pull request:
Implements the Taproot fields for PSBT described in [BIP 371](https://github.com/bitcoin/bips/blob/master/bip-0371.mediawiki).

achow101 committed:
Move individual KeyOriginInfo de/ser to separate function

To make it easier to de/serialize individual KeyOriginInfo for PSBTs,
separate the actual de/serialization of KeyOriginInfo to its own
function.

This is an additional separation where any length prefix is processed by
the caller.

achow101 committed:
Add TaprootBuilder::GetTreeTuples

GetTreeTuples returns the leaves in DFS order as tuples of depth, leaf
version, and script. This is a representation of the tree that can be
serialized.

achow101 committed:
Add serialization methods to XOnlyPubKey

It is useful to have serialzation methods for XOnlyPubKey. These will
serialize the internal uint256, so it is not prefixed with the length as
CPubKey does.

achow101 committed:
Implement de/ser of PSBT's Taproot fields

achow101 committed:
Fill PSBT Taproot input data to/from SignatureData

achow101 committed:
Fetch key origins for Taproot keys

achow101 committed:
Store TaprootBuilder in SigningProviders instead of TaprootSpendData

TaprootSpendData can be gotten from TaprootBuilder, however for PSBT, we
also need TaprootBuilders directly (for the outputs). So we store the
TaprootBuilder in the FlatSigningProvider and when the TaprootSpendData
is needed, we generate it on the fly using the stored builder.

achow101 committed:
Fill PSBT Taproot output data to/from SignatureData

achow101 committed:
Implement decodepsbt for Taproot fields

achow101 committed:
psbt: Remove non_witness_utxo for segwit v1+

If all inputs are segwit v1+, the non_witness_utxos can be removed.

achow101 committed:
tests: Test taproot fields for PSBT

achow101 committed:
taproot: Use pre-existing signatures if available

Actually use pre-existing signatures in CreateTaprootScriptSig if a
signature is found for the given key and leaf hash.

achow101 committed:
psbt, test: Check for taproot fields in taproot psbt test

mjdietzx commented:
Concept and Approach ACK. I was experimenting with a basic taproot n-of-m multisig using a single OP_CHECKSIGADD based script on top of this branch. Taproot fields in the PSBTs I was creating (and attempting to sign - although I haven't got that working yet) looked good when I inspected them. I did some very light testing of this, mostly sanity checks

Sjors commented:
Needs rebase due to silent merge conflict in `descriptor.cpp`.

achow101 committed:
psbt: Implement merge for Taproot fields

Sjors commented on src/script/sign.h:
```diff
@@ -74,6 +74,7 @@ struct SignatureData {
     std::map<CKeyID, std::pair<CPubKey, KeyOriginInfo>> misc_pubkeys;
     std::vector<unsigned char> taproot_key_path_sig; /// Schnorr signature for key path spending
     std::map<std::pair<XOnlyPubKey, uint256>, std::vector<unsigned char>> taproot_script_sigs; ///< (Partial) schnorr signatures, indexed by XOnlyPubKey and leaf_hash.
+    std::map<XOnlyPubKey, std::pair<std::set<uint256>, KeyOriginInfo>> taproot_misc_pubkeys;
```
57622204c81628f790a5355d2f7197d8ed23ebc1 : worth documenting that "misc" may include the inner key (with an empty set of hashes).

achow101 replied:
Added some docs for this

Sjors commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1081,6 +1081,43 @@ static RPCHelpMan decodepsbt()
                                 {
                                     {RPCResult::Type::STR_HEX, "", "hex-encoded witness data (if any)"},
                                 }},
+                                {RPCResult::Type::STR_HEX, "taproot_key_path_sig", /* optional */ true, "hex-encoded signature for the Taproot key path spend"},
```
48d5949aa94410b12263df9b7a3cbc4d7cd50b6f: having a fully decoded example in a functional test case would be handy

achow101 replied:
Is there not? There are tests added which check for the presence of the taproot fields in psbts spending taproot inputs.

Sjors replied:
I noticed the presence checking tests, though perhaps checking against the fully decoded result could reveal more subtle issues.

Sjors commented on src/psbt.cpp:
```diff
@@ -357,10 +357,10 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
         input.witness_utxo = utxo;
-        // input.non_witness_utxo = nullptr;
+        // We can remove the non_witness_utxo if and only if there are no non-segwit or segwit v0
+        // inputs in this transaction, so we can deal with this later.
```
833ecb8dec1a10e3bc1b8d5bcaf0b6839277f582 "later" -> "in FillPSBT"?

achow101 replied:
Done

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1894,6 +1894,30 @@ TransactionError CWallet::FillPSBT(PartiallySignedTransaction& psbtx, bool& comp
         }
     }
 
+    // Figure out if any witness_utxos should be dropped
```
833ecb8dec1a10e3bc1b8d5bcaf0b6839277f582: `non_witness_utxos` (also below)

Don't do this for `SIGHASH_SINGLE` and `SIGNHASH_ANYONECANPAY`, since new non-taproot inputs might be added later? Or should whoever adds such an input just add them back in? Maybe add a functional test for this.

achow101 replied:
It only matters for `SIGHASH_ANYONECANPAY`. I've added a check for that.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1894,6 +1894,30 @@ TransactionError CWallet::FillPSBT(PartiallySignedTransaction& psbtx, bool& comp
+            to_drop.clear();
+            break;
+        }
+        to_drop.push_back(i);
```
833ecb8dec1a10e3bc1b8d5bcaf0b6839277f582: maybe only do this if there actually _is_ a `non_witness_utxo`? Otherwise you might as well use a `go_drop` bool and loop over all inputs after this loop. Then you also wouldn't need to track the index.

achow101 replied:
Done

Sjors commented on src/script/sign.cpp:
```diff
@@ -149,6 +149,7 @@ static bool CreateTaprootScriptSig(const BaseSignatureCreator& creator, Signatur
     auto it = sigdata.taproot_script_sigs.find(lookup_key);
     if (it != sigdata.taproot_script_sigs.end()) {
         sig_out = it->second;
+        return true;
```
9b86c58d95539056981c2a04a7345b992b1ce70c: this seems like a bug fix, is there a test you could add that demonstrates this?

achow101 replied:
The existing tests already cover this case (try reverting this commit, the tests will fail).

Sjors replied:
On your latest rebase f7b61d2, when I revert 9b86c58d95539056981c2a04a7345b992b1ce70c, then indeed `wallet_taproot.py` fails. Still, having the fix and test in the same commit would be much clearer. Moving this bit from the test over to this commit does the trick:

```patch
-            res = self.psbt_offline.walletprocesspsbt(psbt)
-            assert(res['complete'])
+            res = self.psbt_offline.walletprocesspsbt(psbt=psbt, finalize=False)
+
+            decoded = self.psbt_offline.decodepsbt(res["psbt"])
```

achow101 replied:
There isn't really a specific test case for this bug; the test was written first and the bug found via the test.

achow101 replied:
I've moved the change that causes the test failure into the previous commit.

Sjors reviewed (commented):
Mostly happy with 58af8917bd80b0a918df1a5d38c4b63beae5306f

sipa commented:
In order to make this more testable, would it make sense to add support for OP_CHECKSIGADD-based multisig to descriptors first?

achow101 commented:
> In order to make this more testable, would it make sense to add support for OP_CHECKSIGADD-based multisig to descriptors first?

I think it is fine without those.

achow101 commented on src/psbt.h:
```diff
@@ -56,6 +65,12 @@ struct PSBTInput
+    std::vector<unsigned char> m_tap_key_sig;
+    std::map<std::pair<XOnlyPubKey, uint256>, std::vector<unsigned char>> m_tap_script_sigs;
+    std::map<std::pair<CScript, int>, std::set<std::vector<unsigned char>, ShortestVectorFirstComparator>> m_tap_scripts;
+    std::map<XOnlyPubKey, std::pair<std::set<uint256>, KeyOriginInfo>> m_tap_keypaths;
```
Done

Sjors reviewed (commented):
utACK bd75271a53054bb1bc7d880d5fbf667be4ab5e41

Having a (draft) PR on top of this that implements `OP_CHECKSIGADD` is certainly not harmful.

achow101 commented on src/script/sign.h:
```diff
@@ -65,6 +65,7 @@ typedef std::pair<CPubKey, std::vector<unsigned char>> SigPair;
 struct SignatureData {
     bool complete = false; ///< Stores whether the scriptSig and scriptWitness are complete
     bool witness = false; ///< Stores whether the input this SigData corresponds to is a witness input
+    TxoutType spk_type = TxoutType::NONSTANDARD; ///< Stores the type of scriptPubKey the input spends
```
Removed

Sjors commented:
re-utACK 57d354a0d7ba99478eab1afc4304202a07ef4bf7

Sjors commented:
re-utACK 8a675a00053db4621f7f27cf22e8d65713dff03b

michaelfolkson commented:
Concept ACK, Approach ACK.

Will hopefully find the time to test with a hardware wallet, HWI (single sig PSBT as Taproot multisig descriptor [not yet merged](https://github.com/bitcoin/bitcoin/pull/24043)) and do a code review/comparison with the BIP.

Would certainly be nice to get this in and PR author has added high priority label. [External projects](https://github.com/cryptoadvance/specter-desktop/issues/1506) are waiting on it.

prusnak commented:
> edit: What hardware wallets can I test this with?

@michaelfolkson Trezor supports Taproot with HWI (master).

prusnak commented:
Shall we try to merge this in time for 23.0? @achow101 @laanwj @MarcoFalke

MarcoFalke commented:
This can't be merged because it needs rebase for several weeks due to a silent conflict

michaelfolkson commented:
Don't know anything about the silent conflict but we are past the [feature freeze](https://github.com/bitcoin/bitcoin/issues/22969) of February 15th too. Hopefully we can get a bunch of new Taproot stuff (e.g. https://github.com/bitcoin/bitcoin/pull/24043) in 24.0.

prusnak commented:
> we are past the [feature freeze](https://github.com/bitcoin/bitcoin/issues/22969) of February 15th too.

ACK

Sjors commented:
This also needs review from more people (I'll re-ACK the rebase once those arrive).

prusnak commented:
> Hopefully we can get a bunch of new Taproot stuff (e.g. #24043) in 24.0.

Let's add this PR (#22558) and PR #24043 to the [`24.0` Milestone](https://github.com/bitcoin/bitcoin/milestone/54) then?

gruve-p commented:
Approach ACK

Empact commented on src/psbt.h:
```diff
@@ -489,6 +569,101 @@ struct PSBTInput
+                    s >> m_tap_key_sig;
+                    if (m_tap_key_sig.size() < 64) {
+                        throw std::ios_base::failure("Input Taproot key path signature is shorter than 64 bytes");
+                    } else if (m_tap_key_sig.size() > 65) {
```
nit: would these sizes be useful extracted as constants?

laanwj replied:
Agree, there's quite some magic values here (also repeated), having constants makes it easier to review.

Empact commented on src/psbt.h:
```diff
@@ -625,6 +837,55 @@ struct PSBTOutput
+                    }
+                    UnserializeFromVector(s, m_tap_internal_key);
+                    break;
+                }
```
nit: some of these blocks are not strictly necessary, e.g. in this case.

laanwj replied:
Why would you say this block isn't necessary?

jb55 commented:
Concept ACK. Testing this now because HWI [isn't signing one of my taproot inputs](https://github.com/bitcoin-core/HWI/issues/585) :(

sipa commented:
@achow101 I wonder if you're still open to modifications of BIP371.

Thinking about later integration with MuSig(2), I think it's somewhat unfortunate that the currently suggested PSBT_IN_TAP_BIP32_DERIVATION field conveys both (a) which keys are participating in which leaves and (b) how those are derived. Post-MuSig2, I expect that often keys will be present that aren't directly BIP32-derived, but are a MuSig* aggregate of BIP32-derived keys. It's not impossible to integrate that on top here, but it'd involve some duplication.

My suggestion would be to have these fields:
* PSBT_IN_TAP_KEY_SIG: unmodified
* PSBT_IN_TAP_SCRIPT_SIG: unmodified
* PSBT_IN_TAP_LEAF_SCRIPT: unmodified
* PSBT_IN_TAP_KEY_TWEAK: combination of PSBT_IN_TAP_INTERNAL_KEY and PSBT_IN_TAP_MERKLE_ROOT; it conveys "the output key is a tweak of this key with this Merkle root".
* PSBT_IN_TAP_KEY_IN_SCRIPT: like PSBT_IN_TAP_BIP32_DERIVATION now, but without the BIP32 path; it just conveys "this key occurs in these leaves".
* PSBT_IN_TAP_BIP32_DERIVATION: like PSBT_IN_TAP_BIP32_DERIVATION, but without the list of leaf hashes. It can be used for (a) the output key directly (b) the internal key (provided in PSBT_IN_TAP_KEY_TWEAK) of (c) a key occurring in a leaf script (provided by PSBT_IN_TAP_KEY_IN_SCRIPT).

achow101 commented:
> I wonder if you're still open to modifications of BIP371.

BIP 371 is already in use in production, notably by Ledger. So I think it might be too late for any modifications.

cc @bigspider

sipa commented:
Ok, no reason to change things in that case.

jb55 commented:
tACK b704884935766748cf533577c1babacfb6d4b5a5

Was able to get mixed taproot keypath + wpkh spend working on HWI with this PR

Haven't looked through the code that much, only tested behavior and I was happy with the result.

bigspider commented:
Thanks @achow101 for the ping.
At this time, since we only support BIP-86 addresses, the only taproot field we rely on is PSBT_IN_TAP_BIP32_DERIVATION, where we assume that the list of hashes is indeed empty (so the expected value is a single byte 0x00, followed by 4 bytes for the fingerprint, followed by the derivation steps.

prusnak commented:
> where we assume that the list of hashes is indeed empty

So it seems that @sipa's suggestion can be implemented in a compatible way that does not break your usage of PSBT_IN_TAP_BIP32_DERIVATION

bigspider commented:
Leaving the `0x00` would not break anything as I understand it.
If desired in order to have a clean standard, I suppose we can have a reasonable upgrade path by letting the next version of the app to skip the `0x00` if present and keeping this loose check for some time.

sanket1729 commented:
> ... I expect that often keys will be present that aren't directly BIP32-derived, but are a MuSig* aggregate of BIP32-derived keys

I was thinking most keys would be BIP32 derived keys of a Musig2 aggregate like `musig2(A,B,C)/*` or BIP32 derived keys of a Musig2 aggregate of BIP32-derived keys, like `musig2(A/*, B/*, C/*)/*`? The case you mention seems like `musgi2(A/*, B/*, C/*)`

michaelfolkson commented:
> @achow101 I wonder if you're still open to modifications of BIP371.

Seems like we're in multi threaded conversation territory here with discussion of possible BIP 371 changes and review of this PR as is. I've opened an issue (#24492) to discuss possible BIP 371 changes in case that helps.

> I'm also happy to take this to the ML if you think this deserves more discussion.

Or alternatively I'll close the issue and possible BIP 371 changes can be taken to the mailing list. I definitely think an email to the mailing list is a good idea once those proposed changes have been solidified.

michaelfolkson commented:
> Seems like we're in multi threaded conversation territory here with discussion of possible BIP 371 changes and review of this PR as is. I've opened an issue (https://github.com/bitcoin/bitcoin/issues/24492) to discuss possible BIP 371 changes in case that helps.

@sipa's [proposed changes](https://github.com/bitcoin/bitcoin/pull/22558#issuecomment-1058426300) won't be made to BIP 371 due to uncertainty over whether other projects have already deployed. Closed the issue.

Sjors commented:
re-utACK 28868805bbf263d75cbd02eb7d510f1e343b979f

guggero commented:
I'm currently testing this PR with/against our PSBT implementation in `lnd`.
I was able to get a BIP86 key spend working :tada: 

But I can't seem to get `bitcoind` to recognize and sign for a key spend path with a merkle root provided.
Perhaps that's not the scope for this PR? If so then I apologize.

According to my question and the answer I got [here](https://bitcoin.stackexchange.com/questions/113489/taproot-signing-with-psbt-how-to-determine-signing-method), it should be enough to set the following fields:

```
    {
      "witness_utxo": {
        "amount": 4.00000000,
        "scriptPubKey": {
          "asm": "1 5a7730b9788468ac1b90baac80005da4965496e4d88bcbac95bd3c83cd906f3a",
          "desc": "addr(bcrt1ptfmnpwtcs352cxush2kgqqza5jt9f9hymz9uhty4h57g8nvsduaq6s9ypp)#q562pd9u",
          "hex": "51205a7730b9788468ac1b90baac80005da4965496e4d88bcbac95bd3c83cd906f3a",
          "address": "bcrt1ptfmnpwtcs352cxush2kgqqza5jt9f9hymz9uhty4h57g8nvsduaq6s9ypp",
          "type": "witness_v1_taproot"
        }
      },
      "taproot_bip32_derivs": [
        {
          "pubkey": "63cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f",
          "master_fingerprint": "00000000",
          "path": "m/86'/0'/0'/0/0",
          "leaf_hashes": [
          ]
        }
      ],
      "taproot_internal_key": "63cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f",
      "taproot_merkle_root": "6c2e4bb01e316abaaee288d69c06cc608cedefd6e1a06813786c4ec51b6e1d38"
    }
```

Full PSBT:
```
cHNidP8BAHECAAAAAT/GF0O3okL4iQV80uYewSbhqa7qWoD2nG489WsPxw9xAAAAAAD/////AoCT3BQAAAAAFgAUKpx5R067Qv191nSkawzjIGJAQ+ri6PoCAAAAABYAFIyIwvqKctH/eWgjlieETO9w55xoAAAAAAABASsAhNcXAAAAACJRIFp3MLl4hGisG5C6rIAAXaSWVJbk2IvLrJW9PIPNkG86IRZjzFLwbgqR4+ZFZjGqBwIh1YzNHtjFtmOdT4h4bc1QDxkAAAAAAFYAAIAAAACAAAAAgAAAAAAAAAAAARcgY8xS8G4KkePmRWYxqgcCIdWMzR7YxbZjnU+IeG3NUA8BGCBsLkuwHjFquq7iiNacBsxgjO3v1uGgaBN4bE7FG24dOAAAIgICZ5Cr1jOOzCBlO+GZ+Y3VSYA26Z/kFrcSYqzM00TKPjQY1G/d61QAAIABAACAAAAAgAEAAAADAAAAAA==
```

But when I use `bitcoin-cli walletprocesspsbt` with that PSBT (the private key is definitely in the wallet, it's the same key I used for the previous BIP86 key spend), then I just get the same PSBT and `"complete": false` back.

Is this me doing something wrong or is this out of scope for this PR?

achow101 commented:
@guggero Are you using a legacy or a descriptor wallet? Is the taproot output script being watched by the wallet (i.e. you imported the descriptor for it)?

guggero commented:
It's a descriptor wallet but I didn't add that specific output descriptor to it. It's the same internal key as was used in the BIP86 test, so the key is there. I assumed the PSBT would be sufficient as the "assembly instruction" for the signature, since the private key is known to the wallet. But I guess I have to learn more about how descriptor wallets work in this context.
I'll try again with the correct descriptor imported and watched by the wallet, that will very likely resolve my issue. Thanks!

Sjors commented:
>  since the private key is known to the wallet.

Afaik we don't sign anything unless we explicitly watch the `scriptPubKey`, which (for descriptor wallets) is only the case if a descriptor has been imported. In addition, that descriptor must have the private key: a descriptor with just the xpub won't work even if we have the corresponding seed. See also #23544.

achow101 committed:
sign: Use sigdata taproot spenddata when signing

The taproot spenddata stored in a sigdata is the combination of data
existing previously (e.g. in a PSBT) and the data stored in a
SigningProvider. In order to use the external data when signing, we need
to be using the sigdata's spenddata.

achow101 committed:
wallet: also search taproot pubkeys in FillPSBT

When filling a PSBT, we search the listed pubkeys in order to determine
whether the current DescriptorScriptPubKeyMan could sign the transaction
even if it is not watching the scripts. With Taproot, the taproot
pubkeys need to be searched as well.

guggero commented:
Thank you very much, this fixed the "key spend with script root" test case.
Is script path signing expected to work as well?

I tried it and got a wrong result. But again, this could be due to me still using the internal and script leaf key from the BIP86 descriptor. But I guess the result is still kind of unexpected. Since I supplied a leaf hash in the BIP32 derivation info and a script with a control block, I don't expect a signature to be put into the final script witness:

```
$ bitcoin-cli walletprocesspsbt cHNidP8BAFICAAAAAV0virKr77xOIRw1FU3mHJUQ6oO7UmoEgIUUsED3ZRiuAQAAAAD/////AYCy5g4AAAAAFgAUy79OrU4hd2M3xvJMufO1pLdEWcUAAAAAAAEBKwCj4REAAAAAIlEgLWIhk/X61lqdbrKkZqR9AsFvT4JFt4lmaYWbtsDTy3ciFcBjzFLwbgqR4+ZFZjGqBwIh1YzNHtjFtmOdT4h4bc1QDyMgY8xS8G4KkePmRWYxqgcCIdWMzR7YxbZjnU+IeG3NUA+swCEWY8xS8G4KkePmRWYxqgcCIdWMzR7YxbZjnU+IeG3NUA85ARuXHYQK/Sy6fb774buzt5a3DU2Va/o8Pz27k/FkM3v/AAAAAFYAAIAAAACAAAAAgAAAAAAAAAAAARcgY8xS8G4KkePmRWYxqgcCIdWMzR7YxbZjnU+IeG3NUA8BGCAblx2ECv0sun2+++G7s7eWtw1NlWv6PD89u5PxZDN7/wAA
{
  "psbt": "cHNidP8BAFICAAAAAV0virKr77xOIRw1FU3mHJUQ6oO7UmoEgIUUsED3ZRiuAQAAAAD/////AYCy5g4AAAAAFgAUy79OrU4hd2M3xvJMufO1pLdEWcUAAAAAAAEBKwCj4REAAAAAIlEgLWIhk/X61lqdbrKkZqR9AsFvT4JFt4lmaYWbtsDTy3cBCEIBQGcgRxlVx6JBG4e3xDCw24Pm8QBFlHpvISfrlUAP1ajvKnh9aAAKCDqGBk5BysaADcNy+KFoBGtRk5GrmhgOwVoAAA==",
  "complete": true
}
```

<details>
  <summary>decoded input PSBT</summary>
<pre>
{
  "tx": {
    "txid": "0faa6cef74415fa47d8ba4c5fbc16ef13ffa6d5bebf5e13f3f07efe255ed54e2",
    "hash": "0faa6cef74415fa47d8ba4c5fbc16ef13ffa6d5bebf5e13f3f07efe255ed54e2",
    "version": 2,
    "size": 82,
    "vsize": 82,
    "weight": 328,
    "locktime": 0,
    "vin": [
      {
        "txid": "ae1865f740b0148580046a52bb83ea10951ce64d15351c214ebcefabb28a2f5d",
        "vout": 1,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967295
      }
    ],
    "vout": [
      {
        "value": 2.50000000,
        "n": 0,
        "scriptPubKey": {
          "asm": "0 cbbf4ead4e21776337c6f24cb9f3b5a4b74459c5",
          "desc": "addr(bcrt1qewl5at2wy9mkxd7x7fxtnua45jm5gkw9fxuurq)#0jsdd5lg",
          "hex": "0014cbbf4ead4e21776337c6f24cb9f3b5a4b74459c5",
          "address": "bcrt1qewl5at2wy9mkxd7x7fxtnua45jm5gkw9fxuurq",
          "type": "witness_v0_keyhash"
        }
      }
    ]
  },
  "global_xpubs": [
  ],
  "psbt_version": 0,
  "proprietary": [
  ],
  "unknown": {
  },
  "inputs": [
    {
      "witness_utxo": {
        "amount": 3.00000000,
        "scriptPubKey": {
          "asm": "1 2d622193f5fad65a9d6eb2a466a47d02c16f4f8245b7896669859bb6c0d3cb77",
          "desc": "addr(bcrt1p943zryl4ltt948twk2jxdfraqtqk7nuzgkmcjenfskdmdsxnedmsfzxf2p)#nyv7jhkl",
          "hex": "51202d622193f5fad65a9d6eb2a466a47d02c16f4f8245b7896669859bb6c0d3cb77",
          "address": "bcrt1p943zryl4ltt948twk2jxdfraqtqk7nuzgkmcjenfskdmdsxnedmsfzxf2p",
          "type": "witness_v1_taproot"
        }
      },
      "taproot_scripts": [
        {
          "script": "2063cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500fac",
          "leaf_ver": 192,
          "control_blocks": [
            "c063cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f"
          ]
        }
      ],
      "taproot_bip32_derivs": [
        {
          "pubkey": "63cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f",
          "master_fingerprint": "00000000",
          "path": "m/86'/0'/0'/0/0",
          "leaf_hashes": [
            "1b971d840afd2cba7dbefbe1bbb3b796b70d4d956bfa3c3f3dbb93f164337bff"
          ]
        }
      ],
      "taproot_internal_key": "63cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f",
      "taproot_merkle_root": "1b971d840afd2cba7dbefbe1bbb3b796b70d4d956bfa3c3f3dbb93f164337bff"
    }
  ],
  "outputs": [
    {
    }
  ],
  "fee": 0.50000000
}
</pre>
</details>

<details>
<summary>decoded output PSBT</summary>
<pre>
{
  "tx": {
    "txid": "0faa6cef74415fa47d8ba4c5fbc16ef13ffa6d5bebf5e13f3f07efe255ed54e2",
    "hash": "0faa6cef74415fa47d8ba4c5fbc16ef13ffa6d5bebf5e13f3f07efe255ed54e2",
    "version": 2,
    "size": 82,
    "vsize": 82,
    "weight": 328,
    "locktime": 0,
    "vin": [
      {
        "txid": "ae1865f740b0148580046a52bb83ea10951ce64d15351c214ebcefabb28a2f5d",
        "vout": 1,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967295
      }
    ],
    "vout": [
      {
        "value": 2.50000000,
        "n": 0,
        "scriptPubKey": {
          "asm": "0 cbbf4ead4e21776337c6f24cb9f3b5a4b74459c5",
          "desc": "addr(bcrt1qewl5at2wy9mkxd7x7fxtnua45jm5gkw9fxuurq)#0jsdd5lg",
          "hex": "0014cbbf4ead4e21776337c6f24cb9f3b5a4b74459c5",
          "address": "bcrt1qewl5at2wy9mkxd7x7fxtnua45jm5gkw9fxuurq",
          "type": "witness_v0_keyhash"
        }
      }
    ]
  },
  "global_xpubs": [
  ],
  "psbt_version": 0,
  "proprietary": [
  ],
  "unknown": {
  },
  "inputs": [
    {
      "witness_utxo": {
        "amount": 3.00000000,
        "scriptPubKey": {
          "asm": "1 2d622193f5fad65a9d6eb2a466a47d02c16f4f8245b7896669859bb6c0d3cb77",
          "desc": "addr(bcrt1p943zryl4ltt948twk2jxdfraqtqk7nuzgkmcjenfskdmdsxnedmsfzxf2p)#nyv7jhkl",
          "hex": "51202d622193f5fad65a9d6eb2a466a47d02c16f4f8245b7896669859bb6c0d3cb77",
          "address": "bcrt1p943zryl4ltt948twk2jxdfraqtqk7nuzgkmcjenfskdmdsxnedmsfzxf2p",
          "type": "witness_v1_taproot"
        }
      },
      "final_scriptwitness": [
        "6720471955c7a2411b87b7c430b0db83e6f10045947a6f2127eb95400fd5a8ef2a787d68000a083a86064e41cac6800dc372f8a168046b519391ab9a180ec15a"
      ]
    }
  ],
  "outputs": [
    {
    }
  ],
  "fee": 0.50000000
}
</pre>
</details>

sipa commented:
@guggero I haven't inspected things in detail, but I suspect that what's going on is that you've simplify provided sufficient information in the PSBT to construct a key-path spend, so that's what it constructs?

achow101 commented:
@sipa is correct. We will always try to create a key path spend if enough information is available to do so. You've provided the internal key and the derivation path for the internal key, even if it also has a leaf hash.

guggero commented:
Ah, that makes a lot of sense, thank you for your answers, @sipa and @achow101 . Without the internal key and merkle root fields I get the expected witness stack :tada: 

<details>
<summary>Here's my full test protocol.</summary>

 1. import descriptor (`importdescriptors '[{"desc":"tr(tprv8ZgxMBicQKsPekctUrTRdca69GwGek7eoSafZ23pNYEsyHctxwWfUjgsq8UKwo3cLDpViLRv84bxtsZevZ6w4ZJqtMMqDZbnVreNZYbe6bu/86'"'"'/0'"'"'/0'"'"'/0/0)#53fkwtjk","timestamp":"now"}]'`)
 2. BIP 86 key spend
     - send coins to `bcrt1pwr5wkkq5jfq8655dk929mar3588qncd0j24csyq22awcjwzh88mq5uhwph` (BIP86 key spend, internal key `63cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f`)
    - create PSBT that references that output
    - sign with walletprocesspsbt `cHNidP8BAHECAAAAAeFJXY+feoKbYaGLcgspwL36Hwl9Ehr5sF4Zy7nj4TpQAAAAAAAAAAAAAo3j+gIAAAAAFgAUD1vsYoxpaHURXVbiu/mxgWarqwiAdNIaAAAAABYAFCqceUdOu0L9fdZ0pGsM4yBiQEPqAAAAAAABASsAZc0dAAAAACJRIHDo61gUkkB9Uo2xVF30caHOCeGvkquIEApXXYk4Vzn2IRZjzFLwbgqR4+ZFZjGqBwIh1YzNHtjFtmOdT4h4bc1QDxkAAAAAAFYAAIAAAACAAAAAgAAAAAAAAAAAAAAA`
 
 3. script root key spend
    - send coins to `bcrt1ptfmnpwtcs352cxush2kgqqza5jt9f9hymz9uhty4h57g8nvsduaq6s9ypp` (script root key spend, internal key `63cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f`, merkle root `6c2e4bb01e316abaaee288d69c06cc608cedefd6e1a06813786c4ec51b6e1d38`)
    - create PSBT that references that output
    - sign with walletprocesspsbt `cHNidP8BAHECAAAAAT/GF0O3okL4iQV80uYewSbhqa7qWoD2nG489WsPxw9xAAAAAAD/////AoCT3BQAAAAAFgAUKpx5R067Qv191nSkawzjIGJAQ+ri6PoCAAAAABYAFIyIwvqKctH/eWgjlieETO9w55xoAAAAAAABASsAhNcXAAAAACJRIFp3MLl4hGisG5C6rIAAXaSWVJbk2IvLrJW9PIPNkG86IRZjzFLwbgqR4+ZFZjGqBwIh1YzNHtjFtmOdT4h4bc1QDxkAAAAAAFYAAIAAAACAAAAAgAAAAAAAAAAAARcgY8xS8G4KkePmRWYxqgcCIdWMzR7YxbZjnU+IeG3NUA8BGCBsLkuwHjFquq7iiNacBsxgjO3v1uGgaBN4bE7FG24dOAAAIgICZ5Cr1jOOzCBlO+GZ+Y3VSYA26Z/kFrcSYqzM00TKPjQY1G/d61QAAIABAACAAAAAgAEAAAADAAAAAA==`
 
 4. script spend
    - send coins to `bcrt1p943zryl4ltt948twk2jxdfraqtqk7nuzgkmcjenfskdmdsxnedmsfzxf2p` (script spend, internal key `63cc52f06e0a91e3e6456631aa070221d58ccd1ed8c5b6639d4f88786dcd500f`, merkle root `1b971d840afd2cba7dbefbe1bbb3b796b70d4d956bfa3c3f3dbb93f164337bff`)
    - create PSBT that references that output
    - sign with walletprocesspsbt `cHNidP8BAFICAAAAAV0virKr77xOIRw1FU3mHJUQ6oO7UmoEgIUUsED3ZRiuAQAAAAD/////AcBg0hEAAAAAFgAUy79OrU4hd2M3xvJMufO1pLdEWcUAAAAAAAEBKwCj4REAAAAAIlEgLWIhk/X61lqdbrKkZqR9AsFvT4JFt4lmaYWbtsDTy3ciFcBjzFLwbgqR4+ZFZjGqBwIh1YzNHtjFtmOdT4h4bc1QDyMgY8xS8G4KkePmRWYxqgcCIdWMzR7YxbZjnU+IeG3NUA+swCEWY8xS8G4KkePmRWYxqgcCIdWMzR7YxbZjnU+IeG3NUA85ARuXHYQK/Sy6fb774buzt5a3DU2Va/o8Pz27k/FkM3v/AAAAAFYAAIAAAACAAAAAgAAAAAAAAAAAAAA=`
 


</details>

tACK d5674f1

S3RK commented:
I was also testing this PR and found one minor issue.
`analyzepsbt` command doesn't correctly determine next role as "signer" for taproot inputs. This is due to the fact that `SignTaproot` doesn't fill `SignatureData::missing_sigs`.

I'm not sure what is the best way to fix this as it's not always possible to know which signatures are missing for taproot inputs. Based on my current understanding, I propose:
1) fill in `SignatureData::missing_sigs`  as accurate as possible based on known information. Avoid providing wrong hints (e.g. when internal key is invalid)
2) remove the check requiring `missing_sigs` to be non empty in `src/node/psbt.cpp:77`. I think the idea was to guide users as to what they need to proceed with the input. But this check is too strict for taproot as we don't always have this information, but still can determine that "signer" is the next role

laanwj commented:
> I was also testing this PR and found one minor issue.
analyzepsbt command doesn't correctly determine next role as "signer" for taproot inputs.

Is resolving this is a blocker for the PR or, is it something better done in a follow-up?

achow101 commented:
> Is resolving this is a blocker for the PR or, is it something better done in a follow-up?

I think it can be done in a followup.

Sjors commented:
> @guggero I've pushed some commits that should fix that problem. Can you try again?

Those two new commits could use a test case.

laanwj commented on src/psbt.h:
```diff
@@ -489,6 +569,101 @@ struct PSBTInput
+                    }
+                    std::vector<unsigned char> script_v;
+                    s >> script_v;
+                    assert(!script_v.empty());
```
Is using an assert here dangerous? I mean, the parsing shouldn't crash on an invalid PSBT.

Sjors replied:
From commit 088e8366765dc7002e9cbcb45654e0709e233cd5. It's also not clear to me why this would be safe.

achow101 replied:
Changed the assert to just a check that `script_v` is not empty, and it will throw if it is.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2180,6 +2180,19 @@ TransactionError DescriptorScriptPubKeyMan::FillPSBT(PartiallySignedTransaction&
+                const XOnlyPubKey& pubkey = pk_pair.first;
+                for (unsigned char prefix : {0x02, 0x03}) {
+                    unsigned char b[33] = {prefix};
+                    std::copy(pubkey.begin(), pubkey.end(), b + 1);
```
`XOnlyPubKey` can only be 32 bytes, so I don't think it matters? It internally contains a uint256.

achow101 committed:
test: Test signing psbts without explicitly having scripts

achow101 commented:
> Those two new commits could use a test case.

Done.

achow101 commented on src/script/standard.cpp:
```diff
@@ -642,3 +642,18 @@ std::optional<std::vector<std::tuple<int, CScript, int>>> InferTaprootTree(const
+    if (m_branch.size()) {
+        const auto& leaves = m_branch[0]->leaves;
+        for (const auto& leaf : leaves) {
+            uint8_t depth = (uint8_t)leaf.merkle_branch.size();
```
Added an assert.

Sjors commented:
`mwallet_taproot.py` CI failure, probably not spurious?

MarcoFalke commented on src/psbt.cpp:
```diff
@@ -177,6 +218,16 @@ void PSBTOutput::FillSignatureData(SignatureData& sigdata) const
+        sigdata.tr_spenddata.internal_key = m_tap_internal_key;
+    }
+    if (!m_tap_tree.IsEmpty()) {
+        TaprootSpendData spenddata = m_tap_tree.GetSpendData();
```
```
 node1 stderr script/standard.cpp:498:53: runtime error: load of value 21, which is not a valid value for type 'bool'
    #0 0x55ba8cce31d5 in TaprootBuilder::GetSpendData() const src/./src/script/standard.cpp:498:53
    #1 0x55ba8cc138e9 in PSBTOutput::FillSignatureData(SignatureData&) const src/./src/psbt.cpp:225:49

SUMMARY: UndefinedBehaviorSanitizer: invalid-bool-load script/standard.cpp:498:53 in 
```
https://cirrus-ci.com/task/5083471289778176?logs=ci#L3013

Sjors replied:
My guess is that it's unitialized because `GetSpendData()` was called before `Finalize()`. We should probably have an assert for that.

achow101 replied:
I've added a call to `Finalize` after parsing, so hopefully it works now.

achow101 replied:
Fixed.

achow101 committed:
Assert that TaprootBuilder is Finalized during GetSpendData

GetSpendData needs to be finalized in order to be used. To avoid future
bugs, assert `!m_output_key.IsNull()` as m_output_key is only set during
Finalize.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2180,6 +2180,19 @@ TransactionError DescriptorScriptPubKeyMan::FillPSBT(PartiallySignedTransaction&
+                    std::copy(pubkey.begin(), pubkey.end(), b + 1);
+                    CPubKey fullpubkey;
+                    fullpubkey.Set(b, b + 33);
+                    std::unique_ptr<FlatSigningProvider> pk_keys = GetSigningProvider(fullpubkey);
```
58a1a0c7a3f813c887d349ea3670584aff3773ad : we could probably use a `GetSigningProvider` specifically for `XOnlyPubKey`, but that can wait.

Sjors commented on src/script/sign.cpp:
```diff
@@ -230,16 +230,16 @@ static bool SignTaproot(const SigningProvider& provider, const BaseSignatureCrea
         KeyOriginInfo info;
-        if (provider.GetKeyOriginByXOnly(spenddata.internal_key, info)) {
-            auto it = sigdata.taproot_misc_pubkeys.find(spenddata.internal_key);
+        if (provider.GetKeyOriginByXOnly(sigdata.tr_spenddata.internal_key, info)) {
```
37f6d29df5405d0cbebbb4368a6dcbc95c8f00e5 : maybe just squash this entire commit into f5e58b6b3dfa1622d7c03f82318ef06eb6c6818c or f7c7be9794d868a91c25a743f00b3fd351589d5f and add a comment along the lines of the commit message?

achow101 replied:
IMO it's a separate enough change that it should be its own commit.

Sjors commented on src/script/standard.cpp:
```diff
@@ -642,3 +643,19 @@ std::optional<std::vector<std::tuple<int, CScript, int>>> InferTaprootTree(const
+    if (m_branch.size()) {
+        const auto& leaves = m_branch[0]->leaves;
+        for (const auto& leaf : leaves) {
+            assert(leaf.merkle_branch.size() < TAPROOT_CONTROL_MAX_NODE_COUNT);
```
4542ba12866c5f9e7bb0da186800f2f5e450bea6: `<= TAPROOT_CONTROL_MAX_NODE_COUNT `? See also `TaprootBuilder::Insert`. In `descriptor.cpp` we check `branches.size() > TAPROOT_CONTROL_MAX_NODE_COUNT`.

achow101 replied:
Fixed.

Sjors reviewed (approved):
utACK d94fffe683e2bffc0417e5bf72c2c78a9166e1e3 modulo `<=` question

achow101 commented on src/psbt.h:
```diff
@@ -618,7 +618,7 @@ struct PSBTInput
                     s >> script_v;
                     if (script_v.empty()) {
-                        throw std::ios_base:failure("Input Taproot leaf script must be at least 1 byte");
+                        throw std::ios_base::failure("Input Taproot leaf script must be at least 1 byte");
```
Done

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -723,5 +723,46 @@ def test_psbt_input_keys(psbt_input, keys):
         )
         assert_equal(psbt2["fee"], psbt3["fee"])
 
+        self.log.info("Test signing inputs that the wallet has keys for but is not watching the scripts")
```
> I assume this works for individual private keys, but not for master keys?

Yes, we're not doing the dumb thing from legacy wallets where the seed key is a legitimate key we can receive at.

> I also seems to work if you have an xpriv and the descriptor has been expanded far enough. Might be good to add a test case for that behaviour too.

Yes, that's the point, the test already covers that? Unless you mean something else.

Sjors replied:
> Yes, we're not doing the dumb thing from legacy wallets where the seed key is a legitimate key we can receive at.

That's not what I mean. I was referering to be able to sign with any key that derives from the seed, assuming a derivation path is given.

achow101 replied:
We will only sign with child keys. Anything that requires deriving other child keys from master keys is not supported.

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -723,5 +723,46 @@ def test_psbt_input_keys(psbt_input, keys):
+            eckey.generate()
+            privkey = bytes_to_wif(eckey.get_bytes())
+
+            desc = descsum_create("tr(50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0,pk({}))".format(eckey.get_pubkey().get_bytes().hex()))
```
From `wallet_taproot.py`, which got it from the BIP. I've refactored that into `test_framework/key.py` so that it can be shared.

mehdialiakbar31 commented:
> Implements the Taproot fields for PSBT described in [BIP 371](https://github.com/bitcoin/bips/blob/master/bip-0371.mediawiki).

Sjors commented on src/psbt.h:
```diff
@@ -624,6 +838,59 @@ struct PSBTOutput
+                        s_tree >> depth;
+                        s_tree >> leaf_ver;
+                        s_tree >> script;
+                        m_tap_tree->Add((int)depth, script, (int)leaf_ver, true /* track */);
```
This is the only place where we call `TaprootBuilder::Add` (in this PR). We're not checking `leaf_ver` here, so with a corrupt PSBT the `leaf_version & ~TAPROOT_LEAF_MASK` assert gets hit.

MarcoFalke commented:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=49260

achow101 commented:
> https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=49260

#25513 fixes that.
